### ISSUE TITLE
Use ID as repository name

### ISSUE THREAD

**Issue by tt** [docker/distribution#1057]
I'm trying to implement a registry that will map the name of a repository to an ID and use that when persisting. The idea is decoupling names from repositories which could eventually allow renames (although there might be other blockers).

I have overridden the "vars.name" in the context which works for computing the right paths for the storage provider. However, [the manifest store verifies the name](https://github.com/docker/distribution/blob/6712e602b037ef898df333d3527ff48e3c19de5b/registry/storage/manifeststore.go#L109-L111). Why does the manifest care about the name of the repository? It seems entirely unnecessary.

I'm considering decorating the storage driver and making it translate paths using regular expressions but that feels brittle (and you might still pull a manifest that declares a different name).

**Comment by tt** [comment]
@stevvooe, you have reviewed some of my pull requests. Can you please comment on this issue?

**Comment by stevvooe** [comment]
@tt The manifest name is verified to ensure that it is being placed in the correct repository. This is the current data model. The correct approach to allow some sort of "aliasing" would be to replace the manifest store that both looks up the aliases and removes the extra validation. At this time, you may be able to replace the validation function with a option when instantiating the manifest store.

**Comment by tt** [comment]
Thank you. [I implemented a `SkipNameVerification` function](https://github.com/docker/distribution/compare/docker:master...tt:allow-skipping-name-verification) next to the `SkipLayerVerification` function already there.

The problem then becomes how to pass this option. I see that `repository.Manifests(...)` accept options but as I don't control the callers (the built-in handlers), I'm not sure how you suggest I pass that option down.

`repository.Manifests(...)` even ignores the options. I'm summoning @richardscothern who put [a comment in that function](https://github.com/docker/distribution/blob/31a448a628b61bc50d4790b49c489b5747ebeca5/registry/client/repository.go#L150) in case he has suggestions for how to do this.

To recap: I'm trying to support renames. The first step towards this is never using the image name in the storage driver but rewriting to an ID that will never change.

**Comment by tt** [comment]
@rsampio suggested I inject this using a repository middleware. That seems to work.

**Comment by RichardScothern** [comment]
@tt : passing manifest options over the wire is not yet specified or implemented.  Do you need this given you are implementing this feature in the registry?

As to answer your question for calling the options, using repository middleware to return a manifest service configured to not verify the name should work.
